### PR TITLE
handle consensus

### PR DESCRIPTION
**Pull request by andrewkmin**
## Summary & Motivation
$title

TLDR: Previously, the `ACTIVITY_STATUS_CONSENSUS_NEEDED` case was blatantly unhandled, resulting in misleading error messages/failure modes.

This change does the following:
- makes changes to the base sdk-browser and sdk-server clients to handle non-terminal activity statuses
- update various client libraries that depend on sdk-browser and sdk-server to allow for awaiting consensus
- adds helpers to retrieve signed transactions + signed messages once consensus has been met on the Turnkey side (i.e. in case of a timeout)
- `@turnkey/viem` now returns viem `BaseError`s that wrap around `TurnkeyActivityError`s. This also implicitly addresses https://github.com/tkhq/sdk/issues/121
- includes new example scripts for the ethers, viem, and solana packages (e.g. `retry.ts` and `consensus.ts`) which exercise this new functionality

The end user experience is such that if there is a policy enforcing consensus, you can configure a poller to await consensus to be met. See screen recordings in the testing section below.

TODO:
- add unit tests (ideally — this requires policies)

## How I Tested These Changes
local examples

Live polling for consensus during Ethers signing scripts:

https://github.com/user-attachments/assets/d788d721-4016-4d1e-a0f9-0e3ae0d5233f

Consensus is not met at first within the polling retry limit, then later met, and the signature is retroactively retrieved + added to a Solana transaction
https://github.com/user-attachments/assets/af7fb6e3-46df-4785-a8ec-e2e7d4d1ff7e

## Did you add a changeset?

If updating one of our packages, you'll likely need to add a changeset to your PR. To do so, run [`pnpm changeset`](https://pnpm.io/using-changesets#adding-new-changesets). `pnpm changeset` will generate a file where you should write a human friendly message about the changes. Note how this ([example](https://github.com/tkhq/sdk/blob/b409cd06790f011bf939adcf0755499b8e7497ae/.changeset/extra-http-exports.md?plain=1#L1)) includes the package name (should be auto added by the command) along with the type of [semver change (major.minor.patch)](https://semver.org/) (which you should set).

These changes will be used at release time to determine what packages to publish and how to bump their version. For more context see [this comment](https://github.com/tkhq/sdk/pull/67#issuecomment-1568838440).

### CODE REVIEW

File: packages/ethers/src/index.ts [hunk 1]
```diff
@@ -128,13 +128,21 @@ export class TurnkeySigner extends AbstractSigner implements ethers.Signer {
...
-      const activity = await this.client.signTransaction({
+      const { activity, signedTransaction } = await this.client.signTransaction({
         signWith: this.signWith,
         type: "TRANSACTION_TYPE_ETHEREUM",
-        unsignedTransaction: message,
+        unsignedTransaction,
       });
 
-      return assertNonNull(activity?.signedTransaction);
+      if (activity.status !== "ACTIVITY_STATUS_COMPLETED") {
```
Comment (zkharit) [inline packages/ethers/src/index.ts:137]: What exactly does this look like in the case where an activity requires consensus? _Should_ an error be thrown?

Comment (andrewkmin) [reply packages/ethers/src/index.ts:145]: Great question — ended up having the default behavior be to poll for 3 seconds, then return an error. At that point, it's up to the end-user to figure out how they want to handle the error. I've included some examples of handling in the various `retry.ts` files

File: examples/with-ethers/src/consensus.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  const turnkeyClient = new TurnkeyServerSDK({
+    apiBaseUrl: process.env.BASE_URL!,
+    apiPrivateKey: process.env.API_PRIVATE_KEY!,
+    apiPublicKey: process.env.API_PUBLIC_KEY!,
+    defaultOrganizationId: process.env.ORGANIZATION_ID!,
+    // The following config is useful in contexts where an activity requires consensus.
+    // By default, if the activity is not initially successful, it will poll a maximum
+    // of 3 times with an interval of 1000 milliseconds.
```
Comment (r-n-o) [inline examples/with-ethers/src/consensus.ts:27]: Nit: 10k milliseconds according to line 32?

File: examples/with-ethers/src/consensus.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+  print("Network:", `${network} (chain ID ${chainId})`);
+  print("Address:", address);
+  print("Balance:", `${ethers.formatEther(balance)} Ether`);
+  print("Transaction count:", `${transactionCount}`);
+
+  // 1. Sign a raw payload (`eth_sign` style)
+  const message = "Hello Turnkey";
+  const signature = await connectedSigner.signMessage(message);
+  const recoveredAddress = ethers.verifyMessage(message, signature);
```
Comment (r-n-o) [inline examples/with-ethers/src/consensus.ts:66]: What makes this example "consensus.ts"? I think consensus vs. not will be driven by the organization setting? (1 vs N root quorum users?)

(basically, it doesn't seem necessary to have separate consensus examples unless I'm missing something 🤔 )

File: packages/ethers/src/index.ts [hunk 4]
```diff
@@ -128,13 +134,23 @@ export class TurnkeySigner extends AbstractSigner implements ethers.Signer {
...
+      );
 
-      return assertNonNull(activity?.signedTransaction);
+      if (activity.status !== "ACTIVITY_STATUS_COMPLETED") {
+        throw new TurnkeyActivityError({
+          message: `Unexpected activity status: ${activity.status}`,
+          activityId: activity.id,
+          activityStatus: activity.status as TActivityStatus,
+        });
+      }
```
Comment (r-n-o) [inline packages/ethers/src/index.ts:151]: I think the key is to throw a _different_ error when the activity is in need of consensus. Something like:
```ts
if (activity.status == "ACTIVITY_STATUS_NEEDS_CONSENSUS") {
    throw new TurnkeyActivityNeedsConsensus({
          message: `Activity requires consensus`,
          activityId: activity.id,
          activityStatus: activity.status as TActivityStatus,
        });
}
```

Then I agree with @andrewkmin, the retry or error behavior can be up to the caller.

File: examples/with-ethers/src/consensus.ts (hunk 3 above)
Comment (andrewkmin) [reply examples/with-ethers/src/consensus.ts]: good call! not that useful. `retry.ts` should be enough to serve the purpose of demonstrating the scenario where one might be waiting for consensus

File: packages/ethers/src/index.ts (hunk 4 above)
Comment (andrewkmin) [reply packages/ethers/src/index.ts]: ended up adding some bespoke errors to viem. the thing is it needs `.walk()` so I ended up extending the `BaseError` class. let me know what you think — if it adds unnecessary complexity, happy to cut it or reduce it down!

File: examples/with-ethers/src/index.ts [hunk 5]
```diff
@@ -25,6 +33,16 @@ async function main() {
...
+    // The following config is useful in contexts where an activity requires consensus.
+    // By default, if the activity is not initially successful, it will poll a maximum
+    // of 3 times with an interval of 10000 milliseconds.
+    //
+    // -----
+    //
+    // activityPoller: {
+    //   intervalMs: 10_000,
+    //   numRetries: 5,
+    // },
```
Comment (r-n-o) [inline examples/with-ethers/src/index.ts:45]: Is this still relevant?

File: packages/ethers/src/index.ts [hunk 6]
```diff
@@ -240,35 +254,34 @@ export class TurnkeySigner extends AbstractSigner implements ethers.Signer {
...
-      }
+      checkActivityStatus({
+        id: activity.id,
+        status: activity.status,
+      });
 
       result = assertNonNull(activity?.result?.signRawPayloadResult);
     } else {
-      result = await this.client.signRawPayload({
+      // const trying = this.client as TurnkeyClient;
```
Comment (r-n-o) [inline packages/ethers/src/index.ts:264]: this comment was probably not meant to be a permanent comment? :)

File: packages/http/src/shared.ts [hunk 7]
```diff
@@ -5,31 +5,187 @@ export type TActivityResponse = definitions["v1ActivityResponse"];
...
+
+    this.name = "TurnkeyActivityConsensusNeededError";
+    this.activityId = activityId ?? undefined;
+    this.activityStatus = activityStatus ?? undefined;
+    this.activityType = activityType ?? undefined;
+    this.cause = cause ?? undefined;
+  }
+}
+
+export function checkActivityStatus(input: {
```
Comment (r-n-o) [inline packages/http/src/shared.ts:60]: `checkActivityStatus` implies it's a stateless innocuous function but I think this is more heavy than it looks: it will throw errors! So maybe `assertActivityCompleted` since that's kind of what it boils down to?

File: packages/http/src/shared.ts [hunk 8]
```diff
@@ -5,31 +5,187 @@ export type TActivityResponse = definitions["v1ActivityResponse"];
...
+export type TSignature = definitions["v1SignRawPayloadResult"];
+export type TSignedTransaction =
+  definitions["v1SignTransactionResult"]["signedTransaction"];
 
 export class TurnkeyActivityError extends Error {
-  activityId: TActivityId | null;
-  activityStatus: TActivityStatus | null;
-  activityType: TActivityType | null;
-  cause: Error | null;
+  activityId: TActivityId | undefined;
```
Comment (r-n-o) [inline packages/http/src/shared.ts:13]: Why the change from `null` to `undefined`?

File: packages/http/src/shared.ts [hunk 9]
```diff
@@ -5,31 +5,187 @@ export type TActivityResponse = definitions["v1ActivityResponse"];
...
+ */
+export function getSignatureFromActivity(activity: TActivity): TSignature {
+  if (
+    ![
+      "ACTIVITY_TYPE_SIGN_RAW_PAYLOAD",
+      "ACTIVITY_TYPE_SIGN_RAW_PAYLOAD_V2",
+    ].includes(activity.type)
+  ) {
+    throw new TurnkeyActivityError({
+      message: `Unexpected activity type: ${activity.type}`,
```
Comment (r-n-o) [inline packages/http/src/shared.ts:118]: the message could be "Cannot get signature from activity type ...etc...". Otherwise it's a bit unclear why it's "unexpected".

File: packages/http/src/shared.ts [hunk 10]
```diff
@@ -5,31 +5,187 @@ export type TActivityResponse = definitions["v1ActivityResponse"];
...
+ * @return signature {r, s, v}
+ */
+export function getSignatureFromActivity(activity: TActivity): TSignature {
+  if (
+    ![
+      "ACTIVITY_TYPE_SIGN_RAW_PAYLOAD",
+      "ACTIVITY_TYPE_SIGN_RAW_PAYLOAD_V2",
+    ].includes(activity.type)
+  ) {
+    throw new TurnkeyActivityError({
```
Comment (r-n-o) [inline packages/http/src/shared.ts:117]: I'm also thinking this should not be called "TurnkeyActivityError". It's likely an SDK-level / user-level issue if this function is called on the wrong activity type. We might need a plain `BadArgumentError`? Otherwise users will be led to believe the issue is within the turnkey activity processing when they get this error.

File: packages/http/src/shared.ts [hunk 11]
```diff
@@ -5,31 +5,187 @@ export type TActivityResponse = definitions["v1ActivityResponse"];
...
+/**
+ * This function is a helper method to easily extract signature strings from a completed signing activity.
+ * Particularly useful for scenarios where a signature requires consensus
+ *
+ * @param activity the signing activity
+ * @return signatures {r, s, v}[]
+ */
+export function getSignaturesFromActivity(activity: TActivity): TSignature[] {
+  if (!["ACTIVITY_TYPE_SIGN_RAW_PAYLOADS"].includes(activity.type)) {
+    throw new TurnkeyActivityError({
```
Comment (r-n-o) [inline packages/http/src/shared.ts:143]: likewise here `UnexpectedInput` or `BadArgument` would capture more closely the type of error here.

File: packages/sdk-browser/scripts/codegen.js [hunk 12]
```diff
@@ -307,43 +309,52 @@ export class TurnkeySDKClientBase {
...
+
+    const pollStatus = async (activityId: string): Promise<TResponseType> => {
+      const pollBody = { activityId };
+      const pollData = await this.getActivity(pollBody) as ActivityResponse;
+
+      if (attempts > maxRetries) {
+        return handleResponse(pollData);
+      }
+
+      attempts += 1;
```
Comment (r-n-o) [inline packages/sdk-browser/scripts/codegen.js:341]: Nice. Finally fixing the "retry forever" behavior! 🏆

File: packages/sdk-browser/src/__types__/base.ts [hunk 13]
```diff
@@ -44,26 +53,23 @@ export class TurnkeyRequestError extends Error {
...
 }
 
 interface BaseSDKClientConfig {
   apiBaseUrl: string;
   organizationId: string;
-  activityPoller?: {
-    duration: number;
-    timeout: number;
-  };
+  activityPoller?: TActivityPollerConfig | undefined;
```
Comment (r-n-o) [inline packages/sdk-browser/src/__types__/base.ts:72]: Should this be considered a breaking change or nah?

File: packages/viem/src/index.ts [hunk 14]
```diff
@@ -10,11 +10,69 @@ import type {
...
+  TurnkeyClient,
+} from "@turnkey/http";
 import { ApiKeyStamper } from "@turnkey/api-key-stamper";
 import type { TurnkeyBrowserClient } from "@turnkey/sdk-browser";
 import type { TurnkeyServerClient } from "@turnkey/sdk-server";
 
+export type TTurnkeyConsensusNeededErrorType = TurnkeyConsensusNeededError & {
+  name: "TurnkeyConsensusNeededError";
+};
+export class TurnkeyConsensusNeededError extends BaseError {
```
Comment (r-n-o) [inline packages/viem/src/index.ts:30]: Instead of defining these errors here, can we pull that into `http`? Is there a way to comply with viem's error interface (implement `walk` is one thing; maybe others?) without importing viem's BaseError directly?

We'd need to implement this interface in our http/shared errors: https://github.com/wevm/viem/blob/408948bb74685e9d562cd39f50768f9b282463fe/src/errors/base.ts#L37-L83

I don't think it'd hurt anything to have our errors implement more methods than they need to, but maybe not something we need right now; feels a bit awkward to be pushed by a particular framework to implement these methods for all Turnkey errors.

File: packages/http/src/shared.ts (hunk 7 above)
Comment (andrewkmin) [reply packages/http/src/shared.ts:60]: def more accurate1

File: examples/with-ethers/src/index.ts (hunk 5 above)
Comment (andrewkmin) [reply examples/with-ethers/src/index.ts:45]: Yep! a bit ambiguous with the numbers though; will change them such that it might be a bit less confusing

File: packages/http/src/shared.ts (hunk 8 above)
Comment (andrewkmin) [reply packages/http/src/shared.ts:13]: I went into this because of type compatibility, and ended up going with undefined for consistency. (functionally, `null == undefined = true`). Specifically, I feel like `param: <Type X> | undefined` better conveys the idea that the absence of the param is okay

File: packages/http/src/shared.ts (hunk 11 above)
Comment (andrewkmin) [reply packages/http/src/shared.ts]: good call

Comment (andrewkmin) [reply packages/http/src/shared.ts]: ended up going with `InvalidArgumentError`

File: packages/sdk-browser/src/__types__/base.ts (hunk 13 above)
Comment (andrewkmin) [reply packages/sdk-browser/src/__types__/base.ts:70]: It could be, but at the same time don't feel like it quite justifies a major version bump. I've explicitly called it out in the changeset!

File: packages/viem/src/index.ts (hunk 14 above)
Comment (andrewkmin) [reply packages/viem/src/index.ts:30]: +1 on the sentiment. I'm down to keep these encapsulated to just Viem for now

File: packages/sdk-browser/src/__types__/base.ts (hunk 13 above)
Comment (r-n-o) [reply packages/sdk-browser/src/__types__/base.ts:70]: Cool! I'm good with that 👍

File: packages/http/src/shared.ts (hunk 8 above)
Comment (r-n-o) [reply packages/http/src/shared.ts:13]: Is `activityId: TActivityId | undefined;` equivalent to `activityId?: TActivityId;` then?

Comment (r-n-o) [reply packages/http/src/shared.ts:13]: If that's the case I'm a definite +1 on the change; simpler to reason about!

File: packages/http/src/shared.ts [hunk 15]
```diff
@@ -5,31 +5,211 @@ export type TActivityResponse = definitions["v1ActivityResponse"];
...
+
+    this.name = "InvalidArgumentError";
+    this.activityId = activityId ?? undefined;
+    this.activityStatus = activityStatus ?? undefined;
+    this.activityType = activityType ?? undefined;
+    this.cause = cause ?? undefined;
+  }
+}
+
+export function assertActivityCompleted(input: {
```
Comment (r-n-o) [inline packages/http/src/shared.ts:84]: Now that I'm seeing this with fresh eyes: the activity ID isn't used anywhere really, we're simply comparing the passed in `status` to `CONSENSUS_NEEDED` and `COMPLETED`. Don't think it's necessarily worth changing this but...a bit of an odd interface given nothing guarantees the passed in status come from the passed in activity ID :/

Maybe it'd be clearer if a `TActivity` was expected and passed in instead?

**Review by r-n-o — APPROVED** [review]
🚀

File: packages/http/src/shared.ts (hunk 15 above)
Comment (andrewkmin) [reply packages/http/src/shared.ts]: Attempting this here: https://github.com/tkhq/sdk/pull/350/commits/aa16d4d5c4e130a8ede3946eca3ec13868ed437d

The drawback of this approach is that it's a bit duplicative -- e.g. we return something like (paraphrasing)

{
  ...
  result: activityData["result"]["activityResult"],
  activity: activityData
  ...
}

So you see the overlap there, but having the `result` up front and center just makes access easier. The reason we needed this in the first place was originally for type-compatibility, but I realized it generally could be helpful to return all activity details

Comment (r-n-o) [reply packages/http/src/shared.ts]: yeah it's definitely not super clear-cut, but I like this new version: `assertActivityCompleted(activity);` reads suuuper nice 😍

**Review by r-n-o — APPROVED** [review]
